plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1083 @@
+from functools import reduce
 from plasmapy.utils.decorators import preserve_signature
-from plasmapy.utils.exceptions import RelativityWarning, RelativityError
+from plasmapy.utils.exceptions import (PlasmaPyWarning,
+                                       RelativityWarning,
+                                       RelativityError)
 from textwrap import dedent
-from typing import (Dict, Union)
+from typing import (Any, Dict, List, Tuple, Union)
+
+try:
+    from astropy.units.equivalencies import Equivalency
+except ImportError:
+    # astropy defined the Equivalency class in v3.2.1
+    class Equivalency:
+        pass
+
+
+class CheckBase:
+    """
+    Base class for 'Check' decorator classes.
+
+    Parameters
+    ----------
+    checks_on_return
+        specified checks on the return of the wrapped function
+
+    **checks
+        specified checks on the input arguments of the wrapped function
+    """
+    def __init__(self, checks_on_return=None, **checks):
+        self._checks = checks
+        if checks_on_return is not None:
+            self._checks['checks_on_return'] = checks_on_return
+
+    @property
+    def checks(self):
+        """
+        Requested checks on the decorated function's input arguments
+        and/or return.
+        """
+        return self._checks
+
+
+class CheckValues(CheckBase):
+    """
+    A decorator class to 'check' -- limit/control -- the values of input
+    arguments to a function, as well as, the function's return.
+
+    Parameters
+    ----------
+    checks_on_return: Dict[str, bool]
+        Specifications for value checks on the return of the function being wrapped.
+        (see `check values`_ for valid specifications)
+
+    **checks: Dict[str, Dict[str, bool]]
+        Specifications for value checks on the input arguments of the function
+        being wrapped.  Each keyword argument in `checks` is the name of a function
+        argument to be checked and the keyword value contains the value check
+        specifications.
+
+        .. _`check values`:
+
+        The value check specifications are defined within a dictionary containing
+        the keys defined below.  If the dictionary is empty or omitting keys,
+        then the default value will be assumed for the missing keys.
+
+        ================ ======= ================================================
+        Key              Type    Description
+        ================ ======= ================================================
+        can_be_negative  `bool`  [DEFAULT `True`] values can be negative
+        can_be_complex   `bool`  [DEFAULT `False`] values can be complex numbers
+        can_be_inf       `bool`  [DEFAULT `True`] values can be :data:`~numpy.inf`
+        can_be_nan       `bool`  [DEFAULT `True`] values can be :data:`~numpy.nan`
+        none_shall_pass  `bool`  [DEFAULT `False`] values can be a python `None`
+        ================ ======= ================================================
+
```

Comment:
I doubt `CheckValues` will have anything *checks* that are not boolean values.  I kept this format since `CheckUnits` does have non-boolean *checks* and I was keeping the docstrings between the two classes consistent.